### PR TITLE
fix: oss list bucket return all records

### PR DESCRIPTION
**Pull request by shuangkun**
when list a bucket over 100 object,  the response can only return 100 records, so wrote a loop to return all records.

Fixes #TODO

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Thank you. Can you add some notes in the code? Is there any reference to the 100 records limit?

**Comment by shuangkun** [comment]
> Thank you. Can you add some notes in the code? Is there any reference to the 100 records limit?

https://help.aliyun.com/zh/oss/user-guide/list-objects-18?spm=a2c4g.11186623.0.0.275f7711UTgi3H
there are some offical examples.
```
    continueToken := ""
    for {
         lsRes, err := bucket.ListObjectsV2(oss.ContinuationToken(continueToken))
         if err != nil {
             HandleError(err)
          }
          // 打印列举结果。默认情况下，一次返回100条记录。
          for _, object := range lsRes.Objects {
              fmt.Println(object.Key, object.Type, object.Size, object.ETag, object.LastModified, object.StorageClass)
          }
          if lsRes.IsTruncated {
              continueToken = lsRes.NextContinuationToken
          } else {
              break
          }
      }
```

**Review by terrytangyuan — COMMENTED** [review]
Should we also bump max-keys to 1000?

**Comment by shuangkun** [comment]
> Should we also bump max-keys to 1000?

Maybe we need a configuration to limit it, because some customers' oss objects are indeed relatively large, exceeding 1000.

File: workflow/artifacts/oss/oss.go [hunk 1]
```diff
@@ -236,12 +236,21 @@ func (ossDriver *ArtifactDriver) ListObjects(artifact *wfv1.Artifact) ([]string,
 			if err != nil {
 				return !isTransientOSSErr(err), err
 			}
-			results, err := bucket.ListObjectsV2(oss.Prefix(artifact.OSS.Key))
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:239]: oss.Prefix is removed accidentally?

Comment (shuangkun) [reply workflow/artifacts/oss/oss.go:239]: i forget it. thanks.

File: workflow/artifacts/oss/oss.go [hunk 2]
```diff
@@ -236,12 +236,21 @@ func (ossDriver *ArtifactDriver) ListObjects(artifact *wfv1.Artifact) ([]string,
...
+				}
+				// Add to files. By default, 100 records are returned at a time.
+				for _, object := range results.Objects {
+					files = append(files, object.Key)
+				}
+				continueToken = results.NextContinuationToken
+				if !results.IsTruncated {
+					break
+
+				}
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:253]: We should check whether it's truncated before accessing next continuation token

File: workflow/artifacts/oss/oss.go [hunk 3]
```diff
@@ -236,12 +236,21 @@ func (ossDriver *ArtifactDriver) ListObjects(artifact *wfv1.Artifact) ([]string,
...
-			}
-			for _, object := range results.Objects {
-				files = append(files, object.Key)
+			continueToken := ""
+			for {
+				results, err := bucket.ListObjectsV2(oss.Prefix(artifact.OSS.Key), oss.ContinuationToken(continueToken))
+				if err != nil {
+					return !isTransientOSSErr(err), err
+				}
+				// Add to files. By default, 100 records are returned at a time.
```
Comment (terrytangyuan) [inline workflow/artifacts/oss/oss.go:245]: Add a link to https://help.aliyun.com/zh/oss/user-guide/list-objects-18

**Comment by terrytangyuan** [comment]
Can you address my comments and then run some tests if you have OSS available?

File: workflow/artifacts/oss/oss.go (hunk 3 above)
Comment (shuangkun) [reply workflow/artifacts/oss/oss.go]: ok.

File: workflow/artifacts/oss/oss.go (hunk 2 above)
Comment (shuangkun) [reply workflow/artifacts/oss/oss.go]: Yes.

**Comment by terrytangyuan** [comment]
Can you test it against actual OSS environment and see if everything is working correctly?

**Comment by shuangkun** [comment]
> Can you test it against actual OSS environment and see if everything is working correctly?

OK. I'll run some tests later.

**Comment by shuangkun** [comment]
> Can you test it against actual OSS environment and see if everything is working correctly?

I made a test use the file. It works well.
After fixing this problem, more than 100 steps can be triggered when the directory to be listed has more than 100 files.（There were only 100 before）

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: data-transformations-
  labels:
    workflows.argoproj.io/test: "true"
  annotations:
    workflows.argoproj.io/description: |
      This workflow demonstrates using a data template to list in an S3 bucket
      and then process those log files.
    workflows.argoproj.io/version: ">= 3.1.0"
spec:
  entrypoint: data-transformations
  templates:
    - name: data-transformations
      steps:
        - - name: list-log-files
            template: list-log-files
        - - name: process-logs
            template: process-logs
            withParam: "{{steps.list-log-files.outputs.result}}"
            arguments:
              artifacts:
                - name: file
                  oss:
                    endpoint: http://oss-cn-hangzhou.aliyuncs.com
                    bucket: bucket-workflow-artifect
                    accessKeySecret:
                      name: bucket-workflow-artifect-credentials
                      key: accessKey
                    secretKeySecret:
                      name: bucket-workflow-artifect-credentials
                      key: secretKey
                    key: "{{item}}"

              parameters:
                - name: file-name
                  value: "{{item}}"

    - name: list-log-files
      data:
        source:
          artifactPaths:
            name: test-bucket
            oss:
              endpoint: http://oss-cn-hangzhou.aliyuncs.com
              bucket: bucket-workflow-artifect
              key: artifact-passing-wsl88 # this is path in the bucket
              accessKeySecret:
                name: bucket-workflow-artifect-credentials
                key: accessKey
              secretKeySecret:
                name: bucket-workflow-artifect-credentials
                key: secretKey

        transformation:
          - expression: "filter(data, {# startsWith \"arti\"})"
      outputs:
        artifacts:
          - name: file
            path: /tmp

    - name: process-logs
      inputs:
        parameters:
          - name: file-name
        artifacts:
          - name: file
            path: /tmp
      container:
        image:  argoproj/argosay:v2
        command: [ sh, -c ]
        args:
          - |
            echo {{inputs.parameters.file-name}}
            head /tmp
```

**Review by terrytangyuan — APPROVED** [review]
Thanks!

**Pull request merged**